### PR TITLE
feat(RELEASE-978): add image reference check in apply-mapping

### PR DESCRIPTION
**Pull request by mmalina**
containerImage values are required to use a sha reference. A check was added for this. The task will fail if one or more images use an invalid reference.

### CODE REVIEW

File: tasks/apply-mapping/apply-mapping.yaml [hunk 1]
```diff
@@ -46,6 +46,23 @@ spec:
...
         fi
 
+        # images are required to use sha reference - check this
+        CHECK_RESULT=success
+        NUM_COMPONENTS=$(jq '.components | length' "$SNAPSHOT_SPEC_FILE")
+        for ((i=0; i < NUM_COMPONENTS; i++)); do
+          COMPONENT=$(jq -c ".components[$i]" "$SNAPSHOT_SPEC_FILE")
+          NAME=$(jq -r '.name' <<< "$COMPONENT")
+          IMAGE_REF=$(jq -r '.containerImage' <<< "$COMPONENT")
+          if ! [[ "$IMAGE_REF" =~ ^[^:]+@sha256:[0-9a-f]+$ ]] ; then
```
Comment (johnbieren) [inline tasks/apply-mapping/apply-mapping.yaml]: Would it not make more sense to just perform this check in the loop on line 134 instead of looping through the components twice?

File: tasks/apply-mapping/tests/test-apply-mapping-staged-files-expansion.yaml [hunk 2]
```diff
@@ -6,8 +6,8 @@ metadata:
 spec:
   description: |
     Run the apply-mapping task with a snapshot.spec json and a custom mapping provided in
-    the data file with tags per component and verify that the resulting json
-    contains the expected values with tags expanded.
+    the data file with staged files per component and verify that the resulting json
```
Comment (johnbieren) [inline tasks/apply-mapping/tests/test-apply-mapping-staged-files-expansion.yaml:9]: why this change to the description?

Comment (mmalina) [reply tasks/apply-mapping/tests/test-apply-mapping-staged-files-expansion.yaml:9]: Because it seemed obvious this description was copied from another test and the author forgot to change it - there are no tags per component in this test. This test is testing the staged files expansion.

File: tasks/apply-mapping/apply-mapping.yaml (hunk 1 above)
Comment (mmalina) [reply tasks/apply-mapping/apply-mapping.yaml]: I thought it would make sense to have this check at the very beginning, so that we can fail early. I think it also makes the script easier to read, because this block is only doing one thing - checking the image references. But if you think I should move it there, I can do it.

Comment (johnbieren) [reply tasks/apply-mapping/apply-mapping.yaml]: This check will fail the pipeline (it has `exit 1`). So I don't think it matters when it happens in the task, as long as it happens. No external systems are touched in this task (meaning, it isn't like we are pushing to quay or anything) we are just running jq in a workspace that will be immediately removed after as the pipelinerun is going to fail. that is why I think it is better to put it in the other for loop instead of having the same loop twice

File: tasks/apply-mapping/tests/test-apply-mapping-staged-files-expansion.yaml (hunk 2 above)
Comment (johnbieren) [reply tasks/apply-mapping/tests/test-apply-mapping-staged-files-expansion.yaml:9]: Ah! The description tricked me and had me thinking this was actually the tag expansion test. Nice catch

File: tasks/apply-mapping/apply-mapping.yaml (hunk 1 above)
Comment (mmalina) [reply tasks/apply-mapping/apply-mapping.yaml]: OK, now it's in the main loop.

File: tasks/apply-mapping/apply-mapping.yaml [hunk 3]
```diff
@@ -116,6 +116,15 @@ spec:
...
+            NAME=$(jq -r '.name' <<< "$component")
+            IMAGE_REF=$(jq -r '.containerImage' <<< "$component")
+            if ! [[ "$IMAGE_REF" =~ ^[^:]+@sha256:[0-9a-f]+$ ]] ; then
+              echo "Component ${NAME} contains an invalid containerImage value. sha reference is required: ${IMAGE_REF}"
+              exit 1
+            fi
+
             imageTags=$(jq '.tags // []' <<< "$component")
             git_sha=$(jq -r '.source.git.revision' <<< "$component") # this sets the value to "null" if it doesn't exist
             build_sha=$(jq -r '.containerImage' <<< "$component" | cut -d ':' -f 2)
```
Comment (johnbieren) [inline tasks/apply-mapping/apply-mapping.yaml]: Does it make sense to change this to echo $IMAGE_REF | cut -d ':' -f 2

Comment (mmalina) [reply tasks/apply-mapping/apply-mapping.yaml]: Yeah, I thought about it... and wasn't sure if you would like it :D Done.

**Review by johnbieren — APPROVED** [review]
